src/multitask_personalization/rom/models.py

```diff
@@ -0,0 +1,204 @@
+        """Set the reachable points."""
+
+    @abc.abstractmethod
+    def get_reachable_points(self) -> list[NDArray]:
+        """Get the reachable points."""
+
+    @abc.abstractmethod
+    def check_position_reachable(self, position: NDArray) -> bool:
+        """Check if a position is reachable."""
+
+    @abc.abstractmethod
+    def sample_reachable_position(self) -> NDArray:
+        """Sample a reachable position."""
+
+
+class GroundTruthROMModel(ROMModel):
+    """ROM model constructed from ground truth data."""
+
+    def __init__(
+        self,
+        rng: np.random.Generator,
+        subject: int,
+        condition: str,
+        ik_distance_threshold: float = 1e-1,
+    ) -> None:
+        super().__init__()
+        self._rng = rng
+        self._subject = subject
+        self._condition = condition
+        self._ik_distance_threshold = ik_distance_threshold
+        # load array of points from rom/data/rom_points.pkl as self._reachable_joints
+        with open(
+            f"src/multitask_personalization/rom/data/{subject}_{condition}_"
+            + "dense_points.pkl",
+            "rb",
+        ) as f:
+            self._reachable_joints = pickle.load(f)
+        print(
+            f"Loaded {len(self._reachable_joints)} points"
+            + " from {subject}_{condition}_dense_points.pkl"
+        )
+
+    def get_reachable_joints(self) -> NDArray:
+        return self._reachable_joints
+
+    def set_reachable_points(self, reachable_points: list[NDArray]) -> None:
+        self._reachable_points = reachable_points
+        assert len(self._reachable_points) != 0, "No reachable points."
+        self._reachable_kd_tree = KDTree(self._reachable_points)
+        self._upd_reachable = False
+
+    def get_reachable_points(self) -> list[NDArray]:
+        return self._reachable_points
+
+    def check_position_reachable(self, position: NDArray) -> bool:
+        assert not self._upd_reachable, "Must set reachable points first."
+        distance, _ = self._reachable_kd_tree.query(position)
+        return distance < self._ik_distance_threshold
+
+    def sample_reachable_position(self) -> NDArray:
+        assert not self._upd_reachable, "Must set reachable points first."
+        return self._rng.choice(self._reachable_points)
+
+
+class LearnedROMModel(ROMModel):
+    """ROM model learned from data."""
+
+    def __init__(
+        self, rng: np.random.Generator, ik_distance_threshold: float = 1e-1
+    ) -> None:
+        super().__init__()
+        self._rng = rng
+        self._ik_distance_threshold = ik_distance_threshold
+        self._rom_model = MLPROMClassifierTorch(
+            device="cuda" if torch.cuda.is_available() else "cpu"
+        )
+        self._rom_model.load()
+        self._parameter_size = 4
+        self._rom_model_context_parameters = np.array([0.0251, -0.2047, 0.3738, 0.1586])
+        # self._rom_model_context_parameters = np.array(
```

Comment:
remove?

Reply:
These are sample parameters generated with actual functional scores. A large subspace of the parameter space corresponds to models that predicts infeasible for everything, so I believe these would be helpful for determining the action space bounds.

Reply:
got it, thanks! can you write that in a comment in the code?

Reply:
Done, will merge after CI passes

Reply:
Do you have a preference of squash/no squash?

Reply:
thanks! yes, please squash 🦶